src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -442,37 +470,92 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
             return future;
         }
 
+        SpoolMessage spooledMessage;
         try {
-            spool.addMessage(request);
+            spooledMessage = spool.addMessage(request);
             triggerSpooler();
         } catch (InterruptedException | SpoolerStoreException e) {
             logger.atDebug().log("Fail to add publish request to spooler queue", e);
             future.completeExceptionally(e);
             return future;
         }
-        return CompletableFuture.completedFuture(0);
+        return spooledMessage.getFuture();
+    }
+
+    protected void isValidPublishRequest(PublishRequest request) throws SpoolerStoreException {
+        // Payload size should be smaller than MQTT maximum message size
+        int messageSize = request.getPayload().length;
+        if (messageSize > MAXIMUM_SIZE_OF_MQTT_MESSAGE_SIZE) {
+            throw new SpoolerStoreException(String.format("The message size exceeds the maximum size of %d ",
+                    MAXIMUM_SIZE_OF_MQTT_MESSAGE_SIZE));
+        }
+
+        // Topic should not contain wildcard characters
+        String topic = request.getTopic();
+        if (topic.contains("#") || topic.contains("+")) {
+            throw new SpoolerStoreException("The topic of publish request should not contain wildcard "
+                    + "characters of '#' or '+'");
+        }
+
+        String reservedTopicTemplate = "^\\$AWS/rules/\\S+/\\S+";
+        String prefixOfReservedTopic = "^\\$AWS/rules/\\S+?/";
```

Comment:
I'm pretty sure that `$aws` is case insensitive

Reply:
Yes, I will update this.